File: src/analyzers/cognitive.ts
import type { Analyzer, AnalyzerResult, AnalyzerDetail } from './types.js';
import type { CodeBenchConfig } from '../config/index.js';
import type { ParsedFile } from '../parsers/index.js';

/**
 * Compute cognitive complexity using a simplified Sonar model.
 * Increments for control flow breaks, with nesting depth as multiplier.
 */
function computeCognitiveComplexity(body: string): number {
  let complexity = 0;
  let nestingDepth = 0;
  const lines = body.split('\n');

  for (const line of lines) {
    const trimmed = line.trim();

    // Track nesting via braces (simplified)
    const opens = (trimmed.match(/\{/g) || []).length;
    const closes = (trimmed.match(/\}/g) || []).length;

    // Increment for control flow keywords (with nesting penalty)
    if (/\b(if|else if)\s*\(/.test(trimmed)) {
      complexity += 1 + nestingDepth;
    } else if (/\belse\s*\{/.test(trimmed)) {
      complexity += 1;
    }

    if (/\b(for|while)\s*\(/.test(trimmed)) {
      complexity += 1 + nestingDepth;
    }

    if (/\bdo\s*\{/.test(trimmed)) {
      complexity += 1 + nestingDepth;
    }

    if (/\bswitch\s*\(/.test(trimmed)) {
      complexity += 1 + nestingDepth;
    }

    if (/\bcatch\s*\(/.test(trimmed)) {
      complexity += 1 + nestingDepth;
    }

    // Logical operators add flat increment
    const logicalOps = (trimmed.match(/&&|\|\||\?\?/g) || []).length;
    complexity += logicalOps;

    // Ternary
    if (/\?[^?:]*:/.test(trimmed)) {
      complexity += 1 + nestingDepth;
    }

    // Recursion detection (calling own function name) - skip for simplicity in v0.1

    nestingDepth += opens - closes;
    if (nestingDepth < 0) nestingDepth = 0;
  }

  return complexity;
}

export const cognitiveAnalyzer: Analyzer = {
  name: 'cognitive',

  async analyze(
    files: ParsedFile[],
    config: CodeBenchConfig,
  ): Promise<AnalyzerResult> {
    const threshold = config.thresholds.max_cognitive;
    const details: AnalyzerDetail[] = [];
    let totalFunctions = 0;
    let functionsWithinThreshold = 0;

    for (const file of files) {
      for (const fn of file.functions) {
        const complexity = computeCognitiveComplexity(fn.body);
        totalFunctions++;

        if (complexity <= threshold) {
          functionsWithinThreshold++;
        } else {
          details.push({
            file: file.relativePath,
            message: `${fn.name}() has cognitive complexity ${complexity} (threshold: ${threshold})`,
            line: fn.startLine,
            value: complexity,
          });
        }
      }
    }

    if (totalFunctions === 0) {
      return {
        metric: 'cognitive',
        score: 100,
        details: [],
        summary: 'No functions found to analyze',
      };
    }

    const score = Math.round((functionsWithinThreshold / totalFunctions) * 100);
    const violations = totalFunctions - functionsWithinThreshold;

    return {
      metric: 'cognitive',
      score,
      details: details.sort((a, b) => (b.value ?? 0) - (a.value ?? 0)),
      summary: `${violations} of ${totalFunctions} functions exceed cognitive complexity threshold of ${threshold}`,
    };
  },
};

File: internal/analyzer/cognitive.go
package analyzer

import (
	"fmt"
	"math"
	"regexp"
	"sort"
	"strings"

	"github.com/codebench/codebench/internal/config"
	"github.com/codebench/codebench/internal/parser"
)

type CognitiveAnalyzer struct{}

var (
	cogIfPattern     = regexp.MustCompile(`\b(if|else if)\b`)
	cogElsePattern   = regexp.MustCompile(`\belse\s*\{`)
	cogLoopPattern   = regexp.MustCompile(`\b(for|while)\b`)
	cogDoPattern     = regexp.MustCompile(`\bdo\s*\{`)
	cogSwitchPattern = regexp.MustCompile(`\bswitch\b`)
	cogCatchPattern  = regexp.MustCompile(`\bcatch\b`)
	cogLogicalPattern = regexp.MustCompile(`&&|\|\||\?\?`)
	cogTernaryPattern = regexp.MustCompile(`\?[^?:]*:`)
	braceOpenPattern  = regexp.MustCompile(`\{`)
	braceClosePattern = regexp.MustCompile(`\}`)
)

func computeCognitiveComplexity(body string) int {
	complexity := 0
	nestingDepth := 0
	lines := strings.Split(body, "\n")

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)

		opens := len(braceOpenPattern.FindAllString(trimmed, -1))
		closes := len(braceClosePattern.FindAllString(trimmed, -1))

		if cogIfPattern.MatchString(trimmed) {
			complexity += 1 + nestingDepth
		} else if cogElsePattern.MatchString(trimmed) {
			complexity += 1
		}

		if cogLoopPattern.MatchString(trimmed) {
			complexity += 1 + nestingDepth
		}

		if cogDoPattern.MatchString(trimmed) {
			complexity += 1 + nestingDepth
		}

		if cogSwitchPattern.MatchString(trimmed) {
			complexity += 1 + nestingDepth
		}

		if cogCatchPattern.MatchString(trimmed) {
			complexity += 1 + nestingDepth
		}

		logicalOps := len(cogLogicalPattern.FindAllString(trimmed, -1))
		complexity += logicalOps

		if cogTernaryPattern.MatchString(trimmed) {
			complexity += 1 + nestingDepth
		}

		nestingDepth += opens - closes
		if nestingDepth < 0 {
			nestingDepth = 0
		}
	}

	return complexity
}

func (a *CognitiveAnalyzer) Name() config.MetricName {
	return config.MetricCognitive
}

func (a *CognitiveAnalyzer) Analyze(files []*parser.ParsedFile, cfg *config.Config, cwd string) (*Result, error) {
	threshold := cfg.Thresholds.MaxCognitive
	var details []Detail
	totalFunctions := 0
	withinThreshold := 0

	for _, file := range files {
		for _, fn := range file.Functions {
			complexity := computeCognitiveComplexity(fn.Body)
			totalFunctions++

			if complexity <= threshold {
				withinThreshold++
			} else {
				details = append(details, Detail{
					File:    file.RelativePath,
					Message: fmt.Sprintf("%s() has cognitive complexity %d (threshold: %d)", fn.Name, complexity, threshold),
					Line:    fn.StartLine,
					Value:   float64(complexity),
					Severity:  ComputeSeverity(float64(complexity), float64(threshold)),
					Category:  "reduce-nesting",
					Threshold: float64(threshold),
				})
			}
		}
	}

	if totalFunctions == 0 {
		return &Result{
			Metric:  config.MetricCognitive,
			Score:   100,
			Summary: "No functions found to analyze",
		}, nil
	}

	score := int(math.Round(float64(withinThreshold) / float64(totalFunctions) * 100))
	violations := totalFunctions - withinThreshold

	sort.Slice(details, func(i, j int) bool {
		return details[i].Value > details[j].Value
	})

	return &Result{
		Metric:  config.MetricCognitive,
		Score:   score,
		Details: details,
		Summary: fmt.Sprintf("%d of %d functions exceed cognitive complexity threshold of %d", violations, totalFunctions, threshold),
	}, nil
}
